google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -2922,6 +2974,38 @@ boolean isValid() {
     }
   }
 
+  /**
+   * Returns a multiplexed session. The method fallbacks to a regular session if {@link
+   * SessionPoolOptions#useMultiplexedSession} is not set.
+   */
+  SessionFuture getMultiplexedSessionWithFallback() throws SpannerException {
+    if (options.getUseMultiplexedSession()) {
+      ISpan span = tracer.getCurrentSpan();
+      span.addAnnotation("Acquiring multiplexed session");
+      try {
+        SessionFuture sessionFuture = getMultiplexedSession();
+        span.addAnnotation("Acquired multiplexed session", "sessionId", sessionFuture.getName());
+        incrementNumSessionsInUse(true);
+        return sessionFuture;
+      } catch (Throwable t) {
+        span.addAnnotation("No multiplexed session available.");
+        throw SpannerExceptionFactory.asSpannerException(t.getCause());
+      }
+    } else {
+      return getSession();
+    }
+  }
+
+  MultiplexedSessionFuture getMultiplexedSession() {
+    try {
+      return currentMultiplexedSessionReference.get().get();
```

Comment:
* `DatabaseClientImpl` is not modified in this PR and also the unit tests at `DatabaseClientImplTest` was not parameterized in this PR to keep the diff small. Should we take them at  https://github.com/googleapis/java-spanner/pull/2994 because that is the PR which swaps the method?
* I'll try to mitigate the getSession() blocking/non-blocking issue in the current PR. There is one other major refactoring for spans - https://github.com/googleapis/java-spanner/pull/3016 and post this we will parameterize most test classes to cover multiplexed session test coverage.

Reply:
As discussed, I'm preferring approach #2. Created a new wrapper interface `SessionFutureWrapper` and a couple of implementation which allow keeping the getSession() method non-blocking.

Reply:
Ack